File: services/discovery/src/discovery/jobrow.py
"""Typed discovery → jobs-row contract (§10), the Python mirror of
@resume/contracts `DiscoveredJob` (jobRow.ts). Kept in LOCKSTEP with the Zod
shape. Each record is validated against this BEFORE upsert, killing the v1 silent
dict-slice drop (store.py:39 `row = {c: record.get(c) for c in COLUMNS}` dropped
any normalizer key not in COLUMNS without warning).
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, ValidationError

CompanyFlag = Literal["dream", "startup", "return-path"]


class DiscoveredJob(BaseModel):
    """Exactly the 13 store.COLUMNS, now typed (matches DiscoveredJob in
    jobRow.ts). `extra='forbid'` is the anti-drop guard: an unexpected key is a
    loud validation error, not a silent drop."""

    model_config = ConfigDict(extra="forbid")

    id: str
    source: str
    company: str
    title: str
    location: str | None
    remote: bool | None
    url: str | None
    posted_at: str | None  # ISO date
    jd_text: str | None
    status: Literal["new", "skipped"]
    skip_reason: str | None
    company_flags: list[CompanyFlag]
    dedupe_key: str


# The DB column order (the INSERT column list). Kept here as the single ordering
# source so store.py builds the row in a stable order.
COLUMNS = list(DiscoveredJob.model_fields.keys())


def validate_record(record: dict) -> dict:
    """Validate a finalized record against DiscoveredJob; returns a clean dict in
    COLUMNS order. Raises ValidationError on any mismatch (caller logs+skips)."""
    job = DiscoveredJob.model_validate(record)
    return {c: getattr(job, c) for c in COLUMNS}


__all__ = ["COLUMNS", "DiscoveredJob", "ValidationError", "validate_record"]

File: packages/contracts/src/jobRow.ts
// §10 Discovery → jobs-row write contract.
//
// Verdict: REDESIGN (typed/shared) — v1's write is an untyped Python dict sliced
// to 13 columns (store.py:39); any normalizer key not in COLUMNS is silently
// dropped. v2 defines the row shape once here (mirrored as a Python TypedDict/
// pydantic model kept in lockstep) and validates each record before insert.
// This is exactly the 13 store.COLUMNS (store.py:8-22), now typed. Insert stays
// ON CONFLICT DO NOTHING (store.py:27).
import { z } from 'zod';
import { CompanyFlag } from './db.js';

export const DiscoveredJob = z
  .object({
    id: z.string(), // gh-<co>-<id> | lever-… | ashby-… | jobspy-<site>-<id>
    source: z.string(), // greenhouse|lever|ashby|jobspy:<site>|manual
    company: z.string(),
    title: z.string(),
    location: z.string().nullable(),
    remote: z.boolean().nullable(),
    url: z.string().nullable(),
    posted_at: z.string().nullable(), // ISO date
    jd_text: z.string().nullable(),
    status: z.enum(['new', 'skipped']), // discovery only ever writes these
    skip_reason: z.string().nullable(), // 'title:<term>' | 'jd:<term>' | null
    company_flags: z.array(CompanyFlag),
    dedupe_key: z.string(), // norm(company)|norm(title)|norm(location)
  })
  .strict();
export type DiscoveredJob = z.infer<typeof DiscoveredJob>;
